File: baker-test/src/test/scala/com/ing/baker/test/scaladsl/BakerAssertTest.scala
package com.ing.baker.test.scaladsl

import java.util.UUID

import akka.actor.ActorSystem
import com.ing.baker.runtime.akka.AkkaBaker
import com.ing.baker.runtime.scaladsl.EventInstance
import com.ing.baker.test.recipe.WebshopBaker._
import com.ing.baker.test.recipe.WebshopRecipe
import com.ing.baker.test.recipe.WebshopRecipe.{ItemsReserved, OrderPlaced}
import com.typesafe.config.ConfigFactory
import org.scalatest.Assertions
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers
import org.slf4j.LoggerFactory

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.util.{Failure, Try}

class BakerAssertTest extends AnyFlatSpec with Matchers {

  private val log = LoggerFactory.getLogger(getClass)

  private def createBakerAssert(async: Boolean = false): BakerAssert = {
    val recipeInstanceId = UUID.randomUUID().toString
    val sensoryEvent = EventInstance.unsafeFrom(OrderPlaced("order-1", "item-1" :: Nil))

    baker.bake(recipeId, recipeInstanceId)
    if (async) {
      Future {
        Thread.sleep(1000)
        baker.fireEvent(recipeInstanceId, sensoryEvent)
      }
    } else {
      baker.fireEvent(recipeInstanceId, sensoryEvent)
    }

    BakerAssert(baker, recipeInstanceId)
  }

  private def assertFail[T](assertion: => T): Unit =
    Try(assertion) match {
      case Failure(t: Throwable) => log.info(t.getMessage)
      case default => fail(s"assertion error is expected but got $default")
    }

  "BakerAssert object" should "be created" in {
    val baker = AkkaBaker(ConfigFactory.load, ActorSystem.apply("test"))
    BakerAssert.apply(baker, "someProcessId")
  }

  "assertEventsFlow" should "work with happy flow" in {
    val bakerAssert = createBakerAssert()
    bakerAssert
      .waitFor(WebshopRecipe.happyFlow)
      .assertEventsFlow(WebshopRecipe.happyFlow)
  }

  "assertEventsFlow" should "work with some delay" in {
    val bakerAssert = createBakerAssert(true)
    bakerAssert
      .waitFor(WebshopRecipe.happyFlow)
      .assertEventsFlow(WebshopRecipe.happyFlow)
  }

  "assertEventsFlow" should "fail on incorrect events" in {
    assertFail(createBakerAssert()
      .waitFor(WebshopRecipe.happyFlow)
      .assertEventsFlow(WebshopRecipe.happyFlow -- classOf[ItemsReserved]))
  }

  "assertIngredient" should "work for isEqual" in {
    val bakerAssert = createBakerAssert()
    bakerAssert
      .waitFor(WebshopRecipe.happyFlow)
      .assertIngredient("orderId").isEqual("order-1")
  }

  "assertIngredient" should "fail for isEqual" in {
    val bakerAssert = createBakerAssert()
    assertFail(bakerAssert.waitFor(WebshopRecipe.happyFlow)
    .assertIngredient("orderId").isEqual("order-2"))
  }

  "assertIngredient" should "work for notNull" in {
    val bakerAssert = createBakerAssert()
    bakerAssert
      .waitFor(WebshopRecipe.happyFlow)
      .assertIngredient("orderId").notNull()
  }

  "assertIngredient" should "work for isNull" in {
    val bakerAssert = createBakerAssert()
    assertFail(bakerAssert.waitFor(WebshopRecipe.happyFlow)
      .assertIngredient("orderId").isNull)
  }

  "assertIngredient" should "work for customAssert" in {
    val bakerAssert = createBakerAssert()
    bakerAssert.waitFor(WebshopRecipe.happyFlow)
      .assertIngredient("orderId").is(value => Assertions.assert(value.as[String] == "order-1"))
  }

  "assertIngredient" should "fail for customAssert" in {
    val bakerAssert = createBakerAssert()
    assertFail(bakerAssert.waitFor(WebshopRecipe.happyFlow)
      .assertIngredient("orderId").is(value => Assertions.assert(value.as[String] == "order-2")))
  }
}

File: core/baker-test/src/test/java/com/ing/baker/test/javadsl/BakerAssertTest.java
package com.ing.baker.test.javadsl;

import akka.actor.ActorSystem;
import com.ing.baker.compiler.RecipeCompiler;
import com.ing.baker.runtime.akka.AkkaBaker;
import com.ing.baker.runtime.javadsl.Baker;
import com.ing.baker.runtime.javadsl.EventInstance;
import com.ing.baker.runtime.javadsl.InteractionInstance;
import com.ing.baker.test.BakerAssert;
import com.ing.baker.test.javadsl.recipe.OrderPlaced;
import com.ing.baker.test.javadsl.recipe.ReserveItemsImpl;
import com.ing.baker.test.javadsl.recipe.WebshopRecipe;
import com.typesafe.config.ConfigFactory;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

public class BakerAssertTest {

    private static final Baker baker = AkkaBaker.java(ConfigFactory.defaultApplication(), ActorSystem.apply("for-java-tests"),
            Collections.singletonList(InteractionInstance.from(new ReserveItemsImpl())));

    private static final String recipeId;

    static {
        try {
            recipeId = baker.addRecipe(RecipeCompiler.compileRecipe(WebshopRecipe.RECIPE), true)
                    .get(10, TimeUnit.SECONDS);
        } catch (Exception e) {
            throw new AssertionError(e);
        }
    }

    private BakerAssert bakerAssert;

    @Before
    public void init() {
        String recipeInstanceId = UUID.randomUUID().toString();

        baker.bake(recipeId, recipeInstanceId);

        bakerAssert = BakerAssert.of(baker, recipeInstanceId);

        baker.fireEvent(recipeInstanceId,
                EventInstance.from(new OrderPlaced("order-1", Arrays.asList("item-1", "item-2", "item-3"))));
    }

    @Test
    public void testHappy() {
        bakerAssert
                .waitFor(WebshopRecipe.HAPPY_FLOW)
                .logEventNames()
                .logIngredients()
                .logVisualState()
                .assertEventsFlow(WebshopRecipe.HAPPY_FLOW);
    }

    @Test(expected = AssertionError.class)
    public void testHappyFail() {

        bakerAssert
                .waitFor(WebshopRecipe.HAPPY_FLOW)
                .assertEventsFlow(WebshopRecipe.HAPPY_FLOW.remove(OrderPlaced.class));
    }

    @Test
    public void testAssertIngredientIsEqual() {
        bakerAssert
                .waitFor(WebshopRecipe.HAPPY_FLOW)
                .assertIngredient("orderId").isEqual("order-1");
    }

    @Test(expected = AssertionError.class)
    public void testAssertIngredientIsEqualFail() {
        bakerAssert
                .waitFor(WebshopRecipe.HAPPY_FLOW)
                .assertIngredient("orderId").isEqual("order-2");
    }

    @Test
    public void testAssertIngredientIsAbsent() {
        bakerAssert
                .waitFor(WebshopRecipe.HAPPY_FLOW)
                .assertIngredient("not-existing").isAbsent();
    }

    @Test
    public void testAssertIngredientCustom() {
        bakerAssert
                .waitFor(WebshopRecipe.HAPPY_FLOW)
                .assertIngredient("orderId").is(val -> Assert.assertEquals("order-1", val.as(String.class)));
    }
}
